File: internal/heartbeat/watchdog.go
package heartbeat

import (
	"sync"
	"time"
)

// Event types emitted by the watchdog.
type Event struct {
	NodeID string
	Type   string // "online", "degraded", "offline"
}

// Watchdog monitors node heartbeats and emits status change events.
type Watchdog struct {
	mu               sync.Mutex
	interval         time.Duration
	degradedAfter    time.Duration
	offlineAfter     time.Duration
	callback         func(Event)
	running          bool
	stopCh           chan struct{}
	registry         HeartbeatRegistry
	intervalUpdated  bool // flag set when interval changes while running
}

// HeartbeatRegistry is the interface the watchdog needs from the cluster registry.
type HeartbeatRegistry interface {
	GetAll() []HeartbeatNode
	UpdateStatus(id string, status string)
}

// HeartbeatNode is a minimal node interface for the watchdog.
type HeartbeatNode struct {
	ID            string
	LastHeartbeat time.Time
	Status        string
}

// NewWatchdog creates a watchdog that checks heartbeat status.
// checkInterval: how often to check
// degradedAfter: time without heartbeat to mark degraded
// offlineAfter: time without heartbeat to mark offline
func NewWatchdog(reg HeartbeatRegistry, checkInterval, degradedAfter, offlineAfter time.Duration, callback func(Event)) *Watchdog {
	return &Watchdog{
		interval:      checkInterval,
		degradedAfter: degradedAfter,
		offlineAfter:  offlineAfter,
		callback:      callback,
		registry:      reg,
	}
}

// Start begins the watchdog loop.
func (w *Watchdog) Start() {
	w.mu.Lock()
	if w.running {
		w.mu.Unlock()
		return
	}
	w.running = true
	w.stopCh = make(chan struct{})
	w.mu.Unlock()

	go func() {
		ticker := time.NewTicker(w.interval)
		defer ticker.Stop()
		for {
			select {
			case <-ticker.C:
				w.check()
				// Check if interval was updated; recreate ticker if so
				w.mu.Lock()
				if w.intervalUpdated {
					w.intervalUpdated = false
					interval := w.interval
					w.mu.Unlock()
					ticker.Stop()
					ticker = time.NewTicker(interval)
				} else {
					w.mu.Unlock()
				}
			case <-w.stopCh:
				return
			}
		}
	}()
}

// Stop halts the watchdog.
func (w *Watchdog) Stop() {
	w.mu.Lock()
	defer w.mu.Unlock()
	if !w.running {
		return
	}
	w.running = false
	close(w.stopCh)
}

// UpdateIntervals dynamically changes the watchdog timing parameters.
func (w *Watchdog) UpdateIntervals(checkInterval, degradedAfter, offlineAfter time.Duration) {
	w.mu.Lock()
	defer w.mu.Unlock()
	if w.running {
		w.intervalUpdated = true
	}
	w.interval = checkInterval
	w.degradedAfter = degradedAfter
	w.offlineAfter = offlineAfter
}

func (w *Watchdog) check() {
	now := time.Now()
	nodes := w.registry.GetAll()
	for _, n := range nodes {
		elapsed := now.Sub(n.LastHeartbeat)
		var newStatus string
		switch {
		case elapsed >= w.offlineAfter:
			newStatus = "offline"
		case elapsed >= w.degradedAfter:
			newStatus = "degraded"
		default:
			newStatus = "online"
		}

		if newStatus != n.Status {
			w.registry.UpdateStatus(n.ID, newStatus)
			if w.callback != nil {
				w.callback(Event{NodeID: n.ID, Type: newStatus})
			}
		}
	}
}

File: hermes_cluster/core/watchdog.py
"""Heartbeat watchdog — monitors node health via heartbeat staleness.

Python port of Go's internal/heartbeat/watchdog.go.
Runs as a background thread, checks heartbeat age periodically,
and emits status change events (online → degraded → offline).

Designed to work with NodeManager's create_watchdog_registry() adapter.
"""

from __future__ import annotations

import logging
import threading
from datetime import datetime
from typing import Callable, List, Optional

logger = logging.getLogger(__name__)


class WatchdogEvent:
    """Event emitted when a node's status changes."""

    __slots__ = ("node_id", "event_type", "timestamp")

    def __init__(self, node_id: str, event_type: str):
        self.node_id = node_id
        self.event_type = event_type  # "online", "degraded", "offline"
        self.timestamp = datetime.utcnow()

    def __repr__(self) -> str:
        return f"WatchdogEvent({self.node_id!r}, {self.event_type!r})"


class HeartbeatNode:
    """Minimal node info needed by the watchdog."""

    __slots__ = ("node_id", "last_heartbeat", "status")

    def __init__(self, node_id: str, last_heartbeat: datetime, status: str):
        self.node_id = node_id
        self.last_heartbeat = last_heartbeat
        self.status = status


class WatchdogRegistry:
    """Interface the watchdog needs from the cluster state."""

    def get_all_heartbeat_nodes(self) -> List[HeartbeatNode]:
        raise NotImplementedError

    def update_node_status(self, node_id: str, status: str) -> None:
        raise NotImplementedError


class Watchdog:
    """Monitors node heartbeats and emits status change events.

    Parameters:
        registry: provides node heartbeat info and status updates
        check_interval: how often to check (seconds)
        degraded_after: seconds without heartbeat to mark degraded
        offline_after: seconds without heartbeat to mark offline
        callback: called with WatchdogEvent on status change
    """

    def __init__(
        self,
        registry: WatchdogRegistry,
        check_interval: float = 5.0,
        degraded_after: float = 15.0,
        offline_after: float = 30.0,
        callback: Optional[Callable[[WatchdogEvent], None]] = None,
    ):
        self._registry = registry
        self._check_interval = check_interval
        self._degraded_after = degraded_after
        self._offline_after = offline_after
        self._callback = callback
        self._running = False
        self._stop_event = threading.Event()
        self._thread: Optional[threading.Thread] = None

    @property
    def is_running(self) -> bool:
        return self._running

    def start(self) -> None:
        """Start the watchdog loop in a background thread."""
        if self._running:
            return
        self._running = True
        self._stop_event.clear()
        self._thread = threading.Thread(
            target=self._loop, daemon=True, name="cluster-watchdog"
        )
        self._thread.start()
        logger.info(
            "watchdog started: check_interval=%.1fs degraded_after=%.1fs offline_after=%.1fs",
            self._check_interval,
            self._degraded_after,
            self._offline_after,
        )

    def stop(self) -> None:
        """Stop the watchdog loop."""
        if not self._running:
            return
        self._running = False
        self._stop_event.set()
        if self._thread and self._thread.is_alive():
            self._thread.join(timeout=2.0)
        logger.info("watchdog stopped")

    def update_intervals(
        self,
        check_interval: float,
        degraded_after: float,
        offline_after: float,
    ) -> None:
        """Dynamically update timing parameters."""
        self._check_interval = check_interval
        self._degraded_after = degraded_after
        self._offline_after = offline_after

    def check_now(self) -> List[WatchdogEvent]:
        """Run a single check cycle and return events. Useful for testing."""
        return self._check()

    def _loop(self) -> None:
        while self._running:
            self._check()
            self._stop_event.wait(timeout=self._check_interval)

    def _check(self) -> List[WatchdogEvent]:
        now = datetime.utcnow()
        nodes = self._registry.get_all_heartbeat_nodes()
        events: List[WatchdogEvent] = []

        for node in nodes:
            elapsed = (now - node.last_heartbeat).total_seconds()
            if elapsed >= self._offline_after:
                new_status = "offline"
            elif elapsed >= self._degraded_after:
                new_status = "degraded"
            else:
                new_status = "online"

            if new_status != node.status:
                self._registry.update_node_status(node.node_id, new_status)
                evt = WatchdogEvent(node.node_id, new_status)
                events.append(evt)
                if self._callback:
                    try:
                        self._callback(evt)
                    except Exception:
                        logger.exception("watchdog callback error for node %s", node.node_id)

        return events
